ws/adapter.go

```diff
@@ -0,0 +1,83 @@
+package ws
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux"
+	broker "github.com/nats-io/go-nats"
+)
+
+var _ Service = (*adapterService)(nil)
+
+// Service contains publish and subscribe methods necessary for
+// message transfer.
+type Service interface {
+	mainflux.MessagePublisher
+
+	// HandleMessage that is received from message broker.
+	HandleMessage(*broker.Msg)
+
+	// AddConnection adds new client ws connection for given client and channel.
+	AddConnection(string, string, *websocket.Conn)
+}
+
+type adapterService struct {
+	pub    mainflux.MessagePublisher
+	conns  map[string]map[string]*websocket.Conn
```

Comment:
It is way simpler to use structs as keys. Check out https://blog.golang.org/go-maps-in-action, section *Key types*.